checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java

```diff
@@ -96,28 +106,374 @@ public LockTransfer createFlowTransferFunction(CFAbstractAnalysis<CFValue, LockS
-        shouldCache = false;
+    class LockQualifierHierarchy extends MultiGraphQualifierHierarchy {
+
+        public LockQualifierHierarchy(MultiGraphFactory f) {
+            super(f, LOCKHELD);
+        }
+        
+
+        boolean isGuardedBy(AnnotationMirror am) {
+            return AnnotationUtils.areSameIgnoringValues(am, GUARDEDBY) ||
+                   AnnotationUtils.areSameIgnoringValues(am, JAVAXGUARDEDBY) ||
+                   AnnotationUtils.areSameIgnoringValues(am, JCIPGUARDEDBY);
+        }
+
+        @Override
+        public boolean isSubtype(AnnotationMirror rhs, AnnotationMirror lhs) {
+
+            boolean lhsIsGuardedBy = isGuardedBy(lhs);
+            boolean rhsIsGuardedBy = isGuardedBy(rhs);
+
+            if (lhsIsGuardedBy && rhsIsGuardedBy) {
+                // Two @GuardedBy annotations are considered subtypes of each other if and only if their values match exactly.
+
+                List<String> lhsValues =
+                    AnnotationUtils.getElementValueArray(lhs, "value", String.class, true);
+                List<String> rhsValues =
+                    AnnotationUtils.getElementValueArray(rhs, "value", String.class, true);
+
+                return rhsValues.containsAll(lhsValues) && lhsValues.containsAll(rhsValues);
+            }
+
+            boolean lhsIsGuardSatisfied = AnnotationUtils.areSameIgnoringValues(lhs, GUARDSATISFIED);
+            boolean rhsIsGuardSatisfied = AnnotationUtils.areSameIgnoringValues(rhs, GUARDSATISFIED);
+
+            if (lhsIsGuardSatisfied && rhsIsGuardSatisfied) {
+                // Two @GuardSatisfied annotations are considered subtypes of each other if and only if their indices match exactly.
+
+                int lhsIndex =
+                        AnnotationUtils.getElementValue(lhs, "value", Integer.class, true);
+                int rhsIndex =
+                    AnnotationUtils.getElementValue(rhs, "value", Integer.class, true);
+
+                return lhsIndex == rhsIndex;
+            }
+
+            // Remove values from @GuardedBy annotations (and use the Checker Framework's GuardedBy annotation, not JCIP's or Javax's)
+            // for further subtype checking. Remove indices from @GuardSatisfied annotations.
+
+            if (lhsIsGuardedBy) {
+                lhs = GUARDEDBY;
+            }
+            else if (lhsIsGuardSatisfied) {
+                lhs = GUARDSATISFIED;
+            }
+
+            if (rhsIsGuardedBy) {
+                rhs = GUARDEDBY;
+            }
+            else if (rhsIsGuardSatisfied) {
+                rhs = GUARDSATISFIED;
+            }
+
+            return super.isSubtype(rhs, lhs);
+        }
+
+        // For caching results of glbs
+        private Map<AnnotationPair, AnnotationMirror> glbs = null;
+
+        // Same contents as in AnnotatedTypeFactory.java
+        @Override
+        public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror a2) {
```

Comment:
I guess the GLB of \@GuardedByInaccessible and \@GuardedBy(something) should be \@GuardedBy(something).  I think it would be better to just write a GLB for the Lock Checker from scratch rather than copying all this code.